Impl `TryFrom` vector for directions and add `InvalidDirectionError`

Jondolf opened the pull request:
# Objective

Implement `TryFrom<Vec2>`/`TryFrom<Vec3>` for direction primitives as considered in #10857.

## Solution

Implement `TryFrom` for the direction primitives.

These are all equivalent:

```rust
let dir2d = Direction2d::try_from(Vec2::new(0.5, 0.5)).unwrap();
let dir2d = Vec2::new(0.5, 0.5).try_into().unwrap(); // (assumes that the type is inferred)
let dir2d = Direction2d::new(Vec2::new(0.5, 0.5)).unwrap();
```

For error cases, an `Err(InvalidDirectionError)` is returned. It contains the type of failure:

```rust
/// An error indicating that a direction is invalid.
#[derive(Debug, PartialEq)]
pub enum InvalidDirectionError {
    /// The length of the direction vector is zero or very close to zero.
    Zero,
    /// The length of the direction vector is `std::f32::INFINITY`.
    Infinite,
    /// The length of the direction vector is `NaN`.
    NaN,
}
```

alice-i-cecile commented on crates/bevy_math/src/primitives/mod.rs:
```diff
@@ -13,6 +13,19 @@ pub trait Primitive2d {}
 
+/// An error indicating that a direction is invalid because it is zero or not finite.
+#[derive(Debug)]
+pub struct InvalidDirectionError;
```
I think this is probably more useful as an enum: the different failure modes are useful to bubble up.

Jondolf replied:
I think we need to do a custom `try_normalize` for this, but I think it's easy since it's just an `.is_finite()` and a comparison against zero

Jondolf replied:
Or alternatively, do the checks again, but only in error cases

Jondolf replied:
I changed it to an enum with `Zero`, `Infinite` and `NaN` variants and added tests to verify that it gives the correct errors.

Jondolf commented:
Is there a reason why the `new` constructor returns an `Option`, or could it also be changed to remove a `Result` with this error? It'd be more descriptive.

devil-ira reviewed (approved):
Looks good to me if the `new` constructor is changed to use this error as well

Jondolf commented:
I changed the `new` constructors to use this error and updated docs accordingly. I also made the docs of the error variants a bit more consistent with each other

rlidwka commented on crates/bevy_math/src/primitives/mod.rs:
```diff
@@ -13,6 +13,26 @@ pub trait Primitive2d {}
+    NaN,
+}
+
+impl std::fmt::Display for InvalidDirectionError {
```
Is there any reason why `thiserror` is not used here?

Jondolf replied:
`bevy_math` doesn't have `thiserror` as a dependency currently. Could add it though, especially if we add more error types
